### PR TITLE
Added LoginServletTests and the servlets package inside test directory.

### PR DESCRIPTION
**Pull request by Bickor**

### CODE REVIEW

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.xml.sax.SAXException;
+
+public class LoginServletTest {
+    private final LocalServiceTestHelper helper =
+            new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
```
Comment (josh8551021) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:36]: Should it be `new LocalUserServiceTestConfig()` in the parentheses?

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        helper.setUp();
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+
+    // Run this test twice to prove we're not leaking any state across tests
+    public void checkIfDatastoreWorks() {
```
Comment (josh8551021) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:49]: Hey, do you need any of this Datastore stuff? It doesn't seem relevant to the testing of the login servlet.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        assertEquals("content type", "application/json", response.getContentType());
+
+        // Turn content to JSon
+        JsonObject jsonObject = turnStringToJson(response.getText());
+
+        assertNotNull("Return Json", response.getText());
+        assertEquals("Json Loggedin", "false", jsonObject.get("LoggedIn").toString());
+
+        // Can't check for the link to be correct, but we can check that it is not null.
+        assertNotNull("Login URL", jsonObject.get("loginURL").toString());
```
Comment (kathleen-lund) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:112]: Would it not work to use a UserService to create the expected login url and check it for correctness against that? If this won't work let me know, just wondering if it's possible.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+        JsonObject jsonObject = turnStringToJson(response.getText());
+
+        assertEquals("LoggedIn", "true", jsonObject.get("LoggedIn").toString());
+
+        // Have to add the extra quotes in the expected because for some reason the JSON Parser adds
+        // them again.
+        assertEquals("Email", "\"example@gmail.com\"", jsonObject.get("Email").toString());
+
+        assertNotNull("Log out url", jsonObject.get("logOutURL").toString());
```
Comment (kathleen-lund) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:142]: Same question about if you can check the URL for correctness by constructing it with UserService?

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 3 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:112]: Yeah you're right! I thought it was more randomly generated but they are actually the same. Corrected.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 1 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:36]: I took it out because it is only needed for datastore testing. Corrected!

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 2 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:49]: We don't took it out. Corrected!

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Test
+    public void testIsNotLoggedIn() {
+        // Check if it works to show the user is not logged in
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(false);
+        UserService userService = UserServiceFactory.getUserService();
+        assertFalse(userService.isUserLoggedIn());
+        // Can't also check if the user is an admin because the method will throw an error since the
+        // user is not even logged in.
+    }
```
Comment (mkmg) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:52]: Are these tests testing any code you wrote? If not, go ahead and remove them (the UserService should have its own tests that verify its behavior).

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        assertEquals("Log out URL", "\"" + userService.createLogoutURL("/") + "\"", jsonObject.get("logOutURL").toString());
+    }
+
+    /**
+     * Turns String into a json object
+     * @param json String to be turned to json
+     * @return JsonObject corresponding to the string.
+     */
+    private static JsonObject turnStringToJson(String json) {
+        JsonParser jsonParser = new JsonParser();
```
Comment (mkmg) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:120]: It doesn't look like this jsonParse variable is used.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        // Can't also check if the user is an admin because the method will throw an error since the
+        // user is not even logged in.
+    }
+
+    @Test
+    public void testDoGetWithoutLogIn() throws IOException, SAXException {
+        // Testing doGet method without having the user logged in.
+        // Result: {"LoggedIn":"false","loginURL":"someURL"}
+
+        ServletRunner sr = new ServletRunner();
```
Comment (tzsebe) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:59]: Hmmm interesting. I would have thought something like this would suffice: https://stackoverflow.com/questions/5434419/how-to-test-my-servlet-using-junit

But This is cool too, if a little thicker than a unit test.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        // The first parameter has to be the same as whatever goes after the '/' in the urlString for
+        // the GetMethodWebRequest so for example if the first parameter is 'restaurant' then the link
+        // would be: 'http://localhost:8080/restaurant' even though the actual servlet may be called
+        // something else.
+        sr.registerServlet("login", LoginServlet.class.getName());
+        ServletUnitClient sc = sr.newClient();
+
+        // This can throw the SAXException.
+        WebRequest request = new GetMethodWebRequest("http://localhost:8080/login");
```
Comment (tzsebe) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:68]: Does this framework not offer a way to run a servlet with a random port (whatever's available at the time)?

What if you're just running the test suite for your entire system, and many servlet unit tests want to run in parallel? Can they all bind to the same port at the same time?

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        JsonObject jsonObject = turnStringToJson(response.getText());
+
+        assertEquals("LoggedIn", "true", jsonObject.get("LoggedIn").toString());
+
+        // Have to add the extra quotes in the expected because for some reason the JSON Parser adds
+        // them again.
+        assertEquals("Email", "\"example@gmail.com\"", jsonObject.get("Email").toString());
+
+        UserService userService = UserServiceFactory.getUserService();
+        assertEquals("Log out URL", "\"" + userService.createLogoutURL("/") + "\"", jsonObject.get("logOutURL").toString());
```
Comment (tzsebe) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:111]: Random question: Why is your capitalization inconsistent? You have "Email" and "LoggedIn", but then "logOutURL" is lowercase. It'd probably be better to use consistent style.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 6 above)
Comment (tzsebe) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:120]: Delete this line.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 9 above)
Comment (tzsebe) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:111]: Hmmm how come there's quotes embedded into your logout url? Is it because that's the redirect URL and is passed as a parameter to the logout url?

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        helper.setUp();
+    }
+
+    @After
+    public void tearDown() {
+        helper.tearDown();
+    }
+
+    @Test
+    public void testIsLoggedIn() {
```
Comment (tzsebe) [inline capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:36]: This test, and the next one, don't instantiate a LoginServlet, so they aren't actually testing the login servlet, and can probably be deleted - unless they are needed to validate that something else is working that isn't specified in the UserService docs.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 5 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:52]: Yeah I wrote them but you're right since it doesn't really test the functionality of my servlet.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 6 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:120]: Corrected!

Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:120]: Done!

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 9 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:111]: For some reason, when I parse the json String in the turnStringToJson() method, it adds an extra set of quotes as part of the string itself. For now this is the quick solution I found.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 10 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:36]: You're right, they are not needed for the functionality of the servlet. Deleted!

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 9 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:111]: Yeah that is weird. Maybe I just got mixed up between capitalization in variables and with Entity names. Corrected!

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 7 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:59]: I can look into using mockito for the future and then transferring these tests to mockito.

File: capstone/src/test/java/com/google/step/servlets/LoginServletTest.java (hunk 8 above)
Comment (Bickor) [reply capstone/src/test/java/com/google/step/servlets/LoginServletTest.java:68]: Yes they should be all able to bind to the same port to my understanding. But I can look to migrating into mockito for the future.
